x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRestoreSourceService.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.ccr.repository;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.component.AbstractLifecycleComponent;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.core.internal.io.IOUtils;
+import org.elasticsearch.index.engine.Engine;
+import org.elasticsearch.index.shard.IndexEventListener;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.store.Store;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.Consumer;
+
+public class CcrRestoreSourceService extends AbstractLifecycleComponent implements IndexEventListener {
+
+    private static final Logger logger = LogManager.getLogger(CcrRestoreSourceService.class);
+
+    private final Map<String, Engine.IndexCommitRef> onGoingRestores = ConcurrentCollections.newConcurrentMap();
+    private final Map<IndexShard, HashSet<String>> sessionsForShard = new HashMap<>();
+    private final CopyOnWriteArrayList<Consumer<String>> openSessionListeners = new CopyOnWriteArrayList<>();
+    private final CopyOnWriteArrayList<Consumer<String>> closeSessionListeners = new CopyOnWriteArrayList<>();
+
+    public CcrRestoreSourceService(Settings settings) {
+        super(settings);
+    }
+
+    // TODO: Need to register with IndicesService
+    @Override
+    public synchronized void beforeIndexShardClosed(ShardId shardId, @Nullable IndexShard indexShard, Settings indexSettings) {
```

Comment:
this is called before the shard is moved to CLOSED state. In order to prevent that new sessions can still come in after that and create an indexcommit ref that might not be immediately cleaned up, I think we should hook into the `afterIndexShardClosed`, which is called after the shard moved to CLOSED. Similarly, we should then add a guard before the call to `acquireSafeIndexCommit` to make sure that the shard is not closed.

Reply:
Made these changes.